mlir/lib/Dialect/MIOpen/Transforms/AlignTiling.cpp

```diff
@@ -376,35 +456,121 @@ template <typename T> struct MILARewritePattern : public OpRewritePattern<T> {
+  // ThreadwiseCopyV2 Conversion in Fusion.
+  //===----------------------------------------------------------------------===//
+
+  struct ThreadwiseCopyV2RewritePattern
+      : public OpRewritePattern<miopen::ThreadwiseCopyV2Op> {
+    using OpRewritePattern<miopen::ThreadwiseCopyV2Op>::OpRewritePattern;
+
+    LogicalResult matchAndRewrite(miopen::ThreadwiseCopyV2Op op,
+                                  PatternRewriter &b) const override {
+      auto loc = op.getLoc();
+      auto ctx = op.getContext();
+      if ((op.dest().getType().dyn_cast<VectorType>() != nullptr) ||
+          (op.source().getType().dyn_cast<VectorType>() != nullptr))
+        return failure();
+
+      Attribute noTransforms = b.getArrayAttr({});
+
+      // Clonned source and dest transforms are finally swapped correctly here.
+      ArrayAttr sourceTransformsOnOp = op.transforms()[1].cast<ArrayAttr>();
+      ArrayAttr destTransformsOnOp = op.transforms()[0].cast<ArrayAttr>();
+      ArrayAttr sourceTransforms, destTransforms;
+      Value source, dest;
+      std::tie(source, sourceTransforms) =
+          miopen::untransform(b, op.source(), sourceTransformsOnOp);
+      std::tie(dest, destTransforms) =
+          miopen::untransform(b, op.dest(), destTransformsOnOp);
+      SmallVector<int64_t, 6> bounds;
+      llvm::transform(
+          op.bounds().getAsRange<IntegerAttr>(), std::back_inserter(bounds),
+          [](const IntegerAttr &v) -> int64_t { return v.getInt(); });
+
+      int64_t dataPerCopy =
+          op->getAttrOfType<IntegerAttr>("data_per_copy").getInt();
+      auto toLoad = op.dest();
+      auto loadType = toLoad.getType().dyn_cast<MemRefType>();
+      auto shape = loadType.cast<ShapedType>().getShape();
+      auto vecType = VectorType::get(dataPerCopy, loadType.getElementType());
+      bounds[shape.size() - 1] /= dataPerCopy;
+
+      Value c0 = b.create<arith::ConstantIndexOp>(loc, 0);
+      ArrayAttr srcLeftOob, srcRightOob;
+      std::tie(srcLeftOob, srcRightOob) =
+          miopen::computeOobFromTransforms(b, sourceTransforms);
+
+      miopen::TransformingForOp copyLoop = b.create<miopen::TransformingForOp>(
+          loc, ArrayRef<ValueRange>{op.sourceCoord(), op.destCoord()},
+          ArrayRef<Attribute>{sourceTransforms, destTransforms}, bounds,
+          /*forceUnroll=*/true, /*useIndexDiffs=*/true);
+      OpBuilder::InsertionGuard guard(b);
+      b.setInsertionPointToStart(copyLoop.getBody());
+      if (dataPerCopy > 1) {
+        Value loaded = b.create<miopen::BufferLoadOp>(
+            loc, vecType, source, srcLeftOob, srcRightOob,
+            copyLoop.getLowerCoords(/*domain=*/0));
+        SmallVector<Value, 6> indicies;
+        for (uint i = 0; i < shape.size() - 1; ++i) {
+          indicies.push_back(c0);
+        }
+        indicies.push_back(copyLoop.getLowerCoords(/*domain=*/1)[0]);
+        b.create<vector::StoreOp>(loc, loaded, dest, indicies);
```

Comment:
Don't case split vector store and memref store, we have `miopen.in_bounds_store` for a reason